segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/InMemoryLog.java

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.server.logs;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.pravega.common.util.BlockingDrainingQueue;
+import io.pravega.segmentstore.server.logs.operations.Operation;
+import lombok.AccessLevel;
+import lombok.Getter;
+
+/**
+ * {@link BlockingDrainingQueue} implementation for {@link Operation}s. Prevents adding {@link Operation}s out of order.
+ */
+public class InMemoryLog extends BlockingDrainingQueue<Operation> {
+    /**
+     * The last sequence number added. This field is only accessed in {@link #addInternal}, which is guaranteed to be
+     * executed while holding the base class' lock, hence no need for extra synchronization here.
+     */
+    @Getter(AccessLevel.PACKAGE)
+    @VisibleForTesting
+    private long lastSequenceNumber = Operation.NO_SEQUENCE_NUMBER;
+
+    /**
+     * See {@link BlockingDrainingQueue#addInternal}.
+     *
+     * @param item The item to include.
+     * @throws OutOfOrderOperationException If item's Sequence Number is out of order.
+     */
+    @Override
+    protected void addInternal(Operation item) {
+        if (this.lastSequenceNumber >= item.getSequenceNumber()) {
+            throw new OutOfOrderOperationException(String.format("Operation '%s' is out of order. Expected sequence number of at least %s.",
+                    item, this.lastSequenceNumber));
+        }
+        super.addInternal(item);
+        this.lastSequenceNumber = item.getSequenceNumber();
```

Comment:
These two operations are not thread safe but this class extends `BlockingDrainingQueue` which calls itself threadsafe.

There are multiple race conditions:
* There could be two threads the first perform the check and then both call `addInternal` even though the second one is before the first violating the invariant above.
* There could be two threads which insert in one order but set lastSequenceNumber in the other order resulting in the wrong sequence number.

Reply:
Look at the javadoc for `lastSequenceNumber`. This method can only be invoked from its superclass which holds a lock while doing so. As such, no further synchronization is needed.

Reply:
Ah, ok. Please add the @guardedby annotation on both this and the BlockingDrainingQueue.addInternal